nbformat/notebooknode.py

```diff
@@ -1,23 +1,50 @@
 """NotebookNode - adding attribute access to dicts"""
 
 from ipython_genutils.ipstruct import Struct
+from collections import Mapping
+
 
 class NotebookNode(Struct):
     """A dict-like node with attribute-access"""
-    pass
+
+    def __setitem__(self, key, value):
+        if isinstance(value, Mapping):
+            value = from_dict(value)
```

Comment:
from_dict always makes a copy, whereas assignment here shouldn't unless necessary. If you want, from_dict could handle that, but I don't think from_dict should be called except on an actual known duct that's not a notebook node.